### PR TITLE
Replace C# ProtocolException and derived exception classes by

### PR DESCRIPTION
**Pull request by bernardnormier**
InvalidDataException and MarshalException.

InvalidDataException reports any and all problems with bytes received by Ice, including various mismatches that may not be an encoding issue.

And MarshalException is now purely about marshaling (serializing/encoding) errors. There is currently only a handful of such errors. For example, a failure to serialize a C# object before sending it "over the wire" as a sequence of bytes.

### CODE REVIEW

File: csharp/src/Ice/WSTransceiver.cs [hunk 1]
```diff
@@ -13,6 +13,23 @@
 
 namespace IceInternal
 {
+
+    // TODO: refactor and eliminate these exceptions
+    public sealed class WSProtocolException : Exception
```
Comment (bentoi) [inline csharp/src/Ice/WSTransceiver.cs]: Sounds like we could also use InvalidDataException here instead?

File: csharp/src/Ice/BZip2.cs [hunk 2]
```diff
@@ -394,7 +394,7 @@ private static string GetBZ2Error(int error)
 
                 if (rc != BzRunOk)
                 {
-                    throw new Ice.CompressionException($"Bzip2 compress run failed {GetBZ2Error(rc)}");
+                    throw new Ice.MarshalException($"bzip2 compresssion failed: {GetBZ2Error(rc)}");
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/BZip2.cs]: `InvalidDataException` is about invalid data (sequence of bytes) we receive. Here the issue is unlikely to be about the source data (it's not compressible?) or the state of the bzip2 object (if we have a bzip2 object), so I think MarshalException ("can't marshal the data") is the most appropriate.

File: csharp/src/Ice/BZip2.cs [hunk 3]
```diff
@@ -405,7 +405,7 @@ private static string GetBZ2Error(int error)
 
                 if (rc != BzStreamEnd)
                 {
-                    throw new Ice.CompressionException($"Bzip2 compress finish failed {GetBZ2Error(rc)}");
+                    throw new Ice.MarshalException($"bzip2 compresssion failed: {GetBZ2Error(rc)}");
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/BZip2.cs]: See above.

File: csharp/src/Ice/WSTransceiver.cs (hunk 1 above)
Comment (bernardnormier) [reply csharp/src/Ice/WSTransceiver.cs]: All these WS exceptions are thrown and caught in this file. I don't think reusing other exceptions is correct. Ideally, we would not use exceptions for this status communication.

File: csharp/src/Ice/Connection.cs [hunk 4]
```diff
@@ -2032,13 +2027,15 @@ private bool Validate(int operation)
                     byte messageType = _readBuffer[8];
                     if (messageType != Ice1Definitions.ValidateConnectionMessage)
                     {
-                        throw new ConnectionNotValidatedException();
+                        throw new InvalidDataException(@$"received ice1 frame with message type `{messageType
+                            }' before receiving frame with ValidateConnectionMessage");
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/Connection.cs]: fixed

File: csharp/src/Ice/InputStream.cs [hunk 5]
```diff
@@ -68,7 +68,8 @@ internal int Pos
             {
                 if (value < 0 || value > _buffer.Count)
                 {
-                    throw new ArgumentOutOfRangeException(nameof(Pos), "The position value is outside the buffer bounds.");
+                    throw new ArgumentOutOfRangeException(nameof(Pos),
+                        "the position value is outside the buffer bounds.");
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/InputStream.cs]: fixed

File: csharp/src/Ice/Protocol.cs [hunk 6]
```diff
@@ -21,14 +21,15 @@ public enum Protocol : byte
...
+        /// NotSupportedException.</summary>
         /// <param name="protocol">The protocol.</param>
         internal static void CheckSupported(this Protocol protocol)
         {
             // For now, we support only ice1
             if (protocol != Protocol.Ice1)
             {
-                throw new UnsupportedProtocolException("", protocol, Protocol.Ice1);
+                throw new System.NotSupportedException(
+                    $"Ice protocol `{protocol.ToString().ToLower()}' is not supported by this Ice runtime");
```
Comment (pepone) [inline csharp/src/Ice/Protocol.cs]: Maybe we should include the runtime version here

```suggestion
                throw new System.NotSupportedException(
                    $"Ice protocol `{protocol.ToString().ToLower()}' is not supported by Ice `{Util.StringVersion()}' runtime");
```

File: csharp/src/Ice/Encoding.cs [hunk 7]
```diff
@@ -107,7 +107,8 @@ internal void CheckSupported()
             // TODO: add 2.0
             if (this != V1_1)
             {
-                throw new UnsupportedEncodingException("", this, V1_1);
+                throw new NotSupportedException(
+                    $"Ice encoding `{ToString()}' is not supported by this Ice runtime");
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/Encoding.cs]: It's equivalent, I find `ToString()` more readable than `this` here.

File: csharp/src/Ice/Ice1Definitions.cs [hunk 8]
```diff
@@ -71,23 +71,28 @@ internal static void CheckHeader(Span<byte> header)
...
-                    $"received ice1 protocol frame with protocol encoding set to {header[2]}.{header[3]}");
+                throw new InvalidDataException(
+                    $"received ice1 protocol frame with protocol encoding set to {BytesToString(header.Slice(2, 2))}");
             }
         }
+
+        private static string BytesToString(Span<byte> bytes)
+        {
+            return BitConverter.ToString(bytes.ToArray());
+        }
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/Ice1Definitions.cs]: Fixed as suggested.

File: csharp/src/Ice/Protocol.cs (hunk 6 above)
Comment (bernardnormier) [reply csharp/src/Ice/Protocol.cs]: Fixed.

**Pull request merged**